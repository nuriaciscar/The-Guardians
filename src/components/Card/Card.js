import { useEffect, useState } from "react";
import { NavLink } from "react-router-dom";
import useLocalApi from "../../hooks/useLocalApi";

const Card = ({ card: { date, image, text, sectionName, articleSubtitle, body, id, placeHolder } }) => {
  const [iconState, setIconState,] = useState(false);
  const { postLocalApi } = useLocalApi();
  const { getLocalApi, localApi, deleteLocalApi } = useLocalApi();
  useEffect(() => {
    getLocalApi();
  }, [getLocalApi]);


  const initialArticleData = {
    sectionName: sectionName,
    imageSource: image,
    articleDate: date,
    articleTitle: text,
    articleSubtitle: articleSubtitle,
    bodyText: body,
    id: id,
    idDelete: id,
  };

  const postNewOnClick = () => {
    if (placeHolder === "homepage" && !iconState) {
      setIconState(!iconState);
    }

    const repeatedNew = localApi.some((element) => element.id === initialArticleData.id);
    if (placeHolder === "homepage" && !repeatedNew) {
      postLocalApi(initialArticleData);
    }
  }

  const deleteOnClick = () => {
    deleteLocalApi(initialArticleData.id);
  }

  useEffect(() => {
    if (placeHolder === "mynews") {
      setIconState(true);
    }
  }, [placeHolder]);

  return (
    <div className="section__card">
      <img
        src={image}
        alt=""
        className={"card__image"}
        width="200"
        height="100"
      />
      <p className="section__text">{text}</p>
      <div className="main__more">
        <NavLink to="/details" className="main__more__read">
          Read More
        </NavLink>
        <div className="main__read-later">
          <img src="/images/bookmark_false.png" alt="icono false" className={iconState ? "notShow" : "show"} width="17" height="17" onClick={postNewOnClick} />
          <img src="/images/bookmark_true.png" alt="icono true" className={iconState ? "show" : "notShow"} width="17" height="17" onClick={deleteOnClick} />
        </div>
      </div>
    </div>
  );
};
export default Card;
