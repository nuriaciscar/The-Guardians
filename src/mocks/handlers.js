import { rest } from "msw";
import { mockArticle, mockArrayApi } from "./handlersCtx";

export const handlers = [
  rest.get(
    "https://api-guardians.herokuapp.com/news",
    async (req, res, ctx) => {
      const resp = res(ctx.json(mockArrayApi));
      return resp;
    }
  ),
];

export const detailPageHandlers = [
  rest.get(
    "https://content.guardianapis.com/football/live/2021/oct/23/crystal-palace-v-newcastle-leeds-v-wolves-and-more-football-clockwatch-live",
    async (req, res, ctx) => {
      const query = req.url.searchParams;
      query.append("api-key", "54d47472-427a-49dc-b6af-d65d241bc415");
      query.append("show-fields", "all");
      const resp = res(ctx.json(mockArticle));
      return resp;
    }
  ),
];
