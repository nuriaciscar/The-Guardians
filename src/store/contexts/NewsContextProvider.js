import NewsContext from "./NewsContext";
import { useReducer } from "react";
import newsReducer from "../reducers/newsReducer";
import newsReducerSport from "../reducers/newsReducerSport";
import newsReducerScience from "../reducers/newsReducerScience";
import newsReducerLifeStyle from "../reducers/newsReducerLifeStyle";
import localApiReducer from "../reducers/localApiReducer";

// falta newsReducer

function NewsContextProvider({ children }) {
  const [news, dispatch] = useReducer(newsReducer, []);
  const [newsSport, dispatchSport] = useReducer(newsReducerSport, []);
  const [newsScience, dispatchScience] = useReducer(newsReducerScience, []);
  const [newsLifeStyle, dispatchLifeStyle] = useReducer(
    newsReducerLifeStyle,
    []
  );
  const [localApi, dispatchLocalApi] = useReducer(localApiReducer, []);

  return (
    <NewsContext.Provider
      value={{
        news,
        dispatch,
        newsSport,
        dispatchSport,
        newsScience,
        dispatchScience,
        newsLifeStyle,
        dispatchLifeStyle,
        localApi,
        dispatchLocalApi,
      }}
    >
      {children}
    </NewsContext.Provider>
  );
}

export default NewsContextProvider;
