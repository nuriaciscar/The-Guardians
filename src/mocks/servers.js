import { setupServer } from "msw/node";
import { detailPageHandlers, handlers } from "./handlers";

export const server = setupServer(...handlers);
export const serverArticleApiGuardian = setupServer(...detailPageHandlers);
