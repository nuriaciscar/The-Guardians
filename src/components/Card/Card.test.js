import { render, screen } from "@testing-library/react";
import Card from "./Card";
import { BrowserRouter as Router } from "react-router-dom";
import { server } from "../../mocks/server";
import NewsContextProvider from "../../store/contexts/NewsContextProvider";

describe("Given a component Card", () => {
  let card;
  beforeAll(() => {
    card = {
      sectionName: "SECTION",
      date: "20-10-1994",
      image: "http://url/",
      text: "ULTIMAS NOTICIAS",
      articleSubtitle: "Que fuerte todo esto, no me lo creo",
      body: "Woooooooooooooooooooooooooooooooooooooooooooooooow",
      id: 1,
      placeHolder: "homepage",
      idHomepage:
        "lifeandstyle/2021/mar/07/how-we-stay-together-its-within-those-storms-that-you-aim-for-the-sun",
    };

    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  describe("When it receives a card", () => {
    test("Then it should render its image, its date, its text, its subtitle, its bod, its id", () => {
      render(
        <NewsContextProvider>
          <Router>
            <Card card={card} />
          </Router>
        </NewsContextProvider>
      );
    });
    // test("Then it should render its picture, its text and an icon", () => {

    //   render(
    //     <Router>
    //       <Card card={card} />
    //     </Router>
    //   );

    //   const image = screen.getByRole("img", "");
    //   const text = screen.getByText(card.text);
    //   const icon = screen.getByText(card.icon);

    //   expect(image).toBeInTheDocument();
    //   expect(text).toBeInTheDocument();
    //   expect(icon).toBeInTheDocument();
    // });
  });
});
