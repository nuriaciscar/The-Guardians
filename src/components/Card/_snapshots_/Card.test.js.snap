exports[`Then it should render a card with his details inside 1`] = `
<div className="section__card">
      <NavLink to={placeHolder === "homepage" ? `/detail/?id=${idHomepage}` : `detailEditPut/?id=${initialArticleData.id}`} className="main__more__read">
        <img
          src={image}
          alt=""
          className={"card__image"}
          width="200"
          height="100"
        />
      </NavLink>
      <p className="section__text">{text} </p>
      <div className="main__more__card">
        <NavLink to={placeHolder === "homepage" ? `/detail/?id=${idHomepage}` : `detailEditPut/?id=${initialArticleData.id}`} className="main__more__read">
          READ MORE
        </NavLink>
        <div className="main__read-later">

          <img src="/images/bookmark_false.png" alt="icono false" className={iconState ? "notShow" : "show"} width="17" height="17" onClick={postNewOnClick} />
          <img src="/images/bookmark_true.png" alt="icono true" className={iconState ? "show" : "notShow"} width="17" height="17" onClick={deleteOnClick} />

        </div>
      </div>
    </div>
`;